### DISCUSSION TITLE
How to make exported FMU has 32-bit Windows binaries in OpenModelica?

### DISCUSSION THREAD

**Discussion by Amuri44** [OpenModelica/OpenModelica#8534] [Q&A] [answered]
I have a model in openModelica and I have exported it as FMU. I tried to import it by a simulation tool. However, this tool failed to load due to 32-bit Window

**Comment by Amuri44** [comment]
Hello, any idea for that?

**Comment by lochel** [comment]
There is a docker option in OMEdit for the FMI export. This should also be available from the scripting interface I guess. I know that @AnHeuermann is using it a bit. Maybe he can comment too.

**Comment by AnHeuermann** [comment, marked as answer]
Something like this should work:
```
loadFile("../MyMOdel.mo"); getErrorString();

buildModelFMU(MyModel, "2.0", "me_cs", platforms={"x86_64-w64-mingw32 docker run docker.openmodelica.org/msyscross-omsimulator:v2.0"});
getErrorString();
```

But maybe use OMEdit and
![image](https://user-images.githubusercontent.com/38031952/151942354-17af1ef6-dcd9-47ca-add7-c5d7399ad908.png)
instead.

Not sure if it works for Win32 or from Windows at the moment. Also you'll need Docker for this.

**Comment by Amuri44** [comment]
> But maybe use OMEdit and ![image](https://user-images.githubusercontent.com/38031952/151942354-17af1ef6-dcd9-47ca-add7-c5d7399ad908.png) instead.
> 
> Not sure if it works for Win32 or from Windows at the moment. Also you'll need Docker for this. 
@AnHeuermann 


How can I add a docker in OMEdit?

**Comment by AnHeuermann** [comment]
You can use the virtualization software [Docker](https://www.docker.com/) from OMEdit, but only if you have it installed on your machine.

To build an FMU from OMEdit inside a docker container, in this case [multiarch/crossbuild](https://hub.docker.com/r/multiarch/crossbuild), you can go to Tools->options->FMI and under `Platforms` you can choose different types of docker platforms. The first argument `i68-w64-mingw32` is the autoconf [target tripple](https://www.gnu.org/software/autoconf/manual/autoconf-2.70/html_node/Specifying-Target-Triplets.html).

![image](https://user-images.githubusercontent.com/38031952/153582986-26f7ebe3-563c-4d5f-bcb3-66a1a12608c1.png)

This works fine under Linux, but has an unfixed issue on Windows (#7458) and won't work at the moment.

**Comment by AnHeuermann** [comment]
The other way to get a 32 bit Windows FMU would of course be to use the 32bit Windows OMEdit version instead of 64bit, but that's not really a good solution.

**Reply by Amuri44** [reply]
@AnHeuermann thank you very much for your reply. So, If I get it right, there are three options:
1- From the script it works fine (also in a 64-bit Wins) as below,
loadFile("../MyMOdel.mo"); getErrorString();
buildModelFMU(MyModel, "2.0", "me_cs", platforms={"x86_64-w64-mingw32 docker run docker.openmodelica.org/msyscross-omsimulator:v2.0"});
getErrorString();`
`
2- From `OMEdit`, it needs docker and it is not working now due an unfixed issue on Windows.
3- Using a 32 bit Windows OMEdit version instead of 64bit.

**Reply by AnHeuermann** [reply]
You'll need Docker for both 1) and 2). Note how the platform is using a docker command `docker run multiarch/crossbuild`.

The easy solution is just to wait for #7458, but this will take some time.

**Reply by Amuri44** [reply]
I reached a problem where I can not proceed easily. I downloaded the `Docker` and it is asked to download `WL2` and this needs Win 10 not Win 11 (that I have).

**Reply by AnHeuermann** [reply]
Windows Subsystem for Linux 2 (WSL2) is supported on Windows 10 and 11. Actually the support on 11 is even better (GUI included).
I can't guide you through the installation process of WSL2 and Docker, but there are plenty of resources out there:

Installing WSL on Windows
https://docs.microsoft.com/en-us/windows/wsl/install

Docker installation on Windows:
https://docs.docker.com/desktop/windows/install/

But note that there are security concerns and that your organization might not give you enough rights to install this kind of virtualization software.

**Reply by Amuri44** [reply]
Thanks for your guide. Noted!
